### PR TITLE
Update CAPI to v1.2.0

### PR DESCRIPTION
**Pull request by jackfrancis**
**What type of PR is this?**

**What this PR does / why we need it**:

This PR updates the capz CI and code dependencies to use v1.2.0 of capi. See:

- https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.2.0

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/config/azure-dev.yaml [hunk 1]
```diff
@@ -32,8 +32,8 @@ providers:
           new: --metrics-addr=:8080
       files:
         - sourcePath: "../data/shared/v1alpha4/metadata.yaml"
-    - name: v1.1.4
-      value: https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.1.4/core-components.yaml
+    - name: v1.2.0-beta.1
```
Comment (sbueringer) [inline test/e2e/config/azure-dev.yaml:35]:
```suggestion
    - name: v1.2.0
```
I think you have to use v1.2.0 instead. clusterctl uses the latest non-prerelease tag (same for CABPK/KCP).
And it then later finds out that that tag (v0.4.7) doesn't have the right contract

Comment (sbueringer) [reply test/e2e/config/azure-dev.yaml:35]: No idea about the image pull issue though, but this solves the clusterctl issue later on

Comment (sbueringer) [reply test/e2e/config/azure-dev.yaml]: Looks like the pull issue was there before: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2425/pull-cluster-api-provider-azure-e2e/1541449105413246976
Maybe something that should be investigated independently (or is it maybe expected?)

Comment (jackfrancis) [reply test/e2e/config/azure-dev.yaml]: thx!, done

File: config/crd/bases/infrastructure.cluster.x-k8s.io_azureclusteridentities.yaml [hunk 2]
```diff
@@ -56,11 +56,11 @@ spec:
                   either a Service Principal password or certificate secret.
                 properties:
                   name:
-                    description: Name is unique within a namespace to reference a
+                    description: name is unique within a namespace to reference a
```
Comment (CecileRobertMichon) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_azureclusteridentities.yaml:59]: if you're curious, this is what changed https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L929

the diff is generated by the k8s.io/api bump

File: test/e2e/config/azure-dev.yaml [hunk 3]
```diff
@@ -92,7 +92,7 @@ providers:
       replacements:
       - old: "imagePullPolicy: Always"
         new: "imagePullPolicy: IfNotPresent"
-    - name: v1.1.99 # next; use manifest from source files
+    - name: v1.2.99 # next; use manifest from source files
```
Comment (CecileRobertMichon) [inline test/e2e/config/azure-dev.yaml:95]: this is a fake CAPZ version, not related to the CAPI version

should actually be v1.4.99 now if we're trying to represent a future patch of the current minor (it's completely arbitrary so it doesn't matter, just needs to be a valid semver so clusterctl doesn't complain)

**Comment by mboersma** [comment]
/retitle ci: test capz with capi v1.2.0

File: azure/scope/cluster_test.go [hunk 4]
```diff
@@ -154,7 +154,6 @@ func TestAPIServerHost(t *testing.T) {
 				Namespace: "default",
 			},
 		}
-		cluster.Default()
```
Comment (CecileRobertMichon) [inline azure/scope/cluster_test.go:157]: what changed here?

**Comment by CecileRobertMichon** [comment]
Let's squash commits?

Did someone check https://github.com/fabriziopandini/cluster-api/blob/main/docs/book/src/developer/providers/v1.1-to-v1.2.md to make sure we did everything on that list?

File: test/e2e/config/azure-dev.yaml (hunk 3 above)
Comment (CecileRobertMichon) [reply test/e2e/config/azure-dev.yaml:95]: this still needs to be fixed @mboersma

**Comment by mboersma** [comment]
> Let's squash commits?

Sure, and I'll rebase off `main`.

> Did someone check https://github.com/fabriziopandini/cluster-api/blob/main/docs/book/src/developer/providers/v1.1-to-v1.2.md to make sure we did everything on that list?

I hadn't, I just did a search-and-replace of versions to update this. I'll go through it–I can see already that we need to update to Go 1.18 to comply.

**Comment by CecileRobertMichon** [comment]
Would be good to add CAPI tests that were recently added to https://github.com/kubernetes-sigs/cluster-api/tree/main/test/e2e to test new v1.2 functionality to our CAPI test suite too: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/capi_test.go

**Comment by mboersma** [comment]
I think in combination with #2487, this PR should cover everything in the [v1.2.0 upgrade doc](https://github.com/fabriziopandini/cluster-api/blob/main/docs/book/src/developer/providers/v1.1-to-v1.2.md).

> Would be good to add CAPI tests that were recently added...

Definitely–I started looking at this but I'm having problems getting `ClusterClassChangesSpecInput` to work currently. Maybe we could add the new tests in a separate PR.

File: azure/scope/cluster_test.go (hunk 4 above)
Comment (jackfrancis) [reply azure/scope/cluster_test.go:157]: See:

- https://github.com/kubernetes-sigs/cluster-api/pull/6225

In particular:

- https://github.com/kubernetes-sigs/cluster-api/pull/6225/files#diff-3dd0f28c0a5df6e110b3304d8ea006b42cd84a5c0a1ef305d3c1dab323924a68L55

File: test/e2e/config/azure-dev.yaml (hunk 3 above)
Comment (jackfrancis) [reply test/e2e/config/azure-dev.yaml]: When updated to 1.4.99 these test failures resulted:

- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2431/pull-cluster-api-provider-azure-e2e/1550212646572134400

I reverted to 1.2.99, there clearly seems to be some significance w/ respect to the way that the cluster-api `metadata.yaml` spec processes this as input.

Comment (jackfrancis) [reply test/e2e/config/azure-dev.yaml]: After further consideration, I think we should do this in a separate PR:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2497

I've advocated that we actually use a "vNext" approach given that that's what main branch is meant to be testing. We can discuss that in the above PR thread.

**Comment by jackfrancis** [comment]
@CecileRobertMichon @mboersma I looked through the provider 1.1-->1.2 doc and identified three things:

1. update to golang 1.18
  - I think we can do this in a separate PR (and I think we can do this either before or after this merges, probably easier after)
2. update slice types for ClusterClass
  - https://github.com/fabriziopandini/cluster-api/blob/main/docs/book/src/developer/providers/v1.1-to-v1.2.md#required-api-changes-for-providers
    - unsure if we need to worry about this in this PR (should we validate this change before merging 1.2 integration at all?) or in a follow-up PR
3. Deprecate `ExperimentalRetryJoin`
  - I've opened an issue for this, we can definitely do this after this PR but prior to releasing 1.5: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2496

**Comment by jackfrancis** [comment]
> @CecileRobertMichon @mboersma I looked through the provider 1.1-->1.2 doc and identified three things:
> 
> 1. update to golang 1.18
> 
> * I think we can do this in a separate PR (and I think we can do this either before or after this merges, probably easier after)
> 
> 2. update slice types for ClusterClass
> 
> * https://github.com/fabriziopandini/cluster-api/blob/main/docs/book/src/developer/providers/v1.1-to-v1.2.md#required-api-changes-for-providers
>   
>   * unsure if we need to worry about this in this PR (should we validate this change before merging 1.2 integration at all?) or in a follow-up PR
> 
> 3. Deprecate `ExperimentalRetryJoin`
> 
> * I've opened an issue for this, we can definitely do this after this PR but prior to releasing 1.5: [Remove UseExperimentalRetryJoin from reference templates #2496](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2496)

Update: golang 1.18 PR merged, and rebased on top of that, so that line item is DONE

File: exp/controllers/azuremanagedcluster_controller.go [hunk 5]
```diff
@@ -86,7 +86,7 @@ func (amcr *AzureManagedClusterReconciler) SetupWithManager(ctx context.Context,
 	// Add a watch on clusterv1.Cluster object for unpause notifications.
 	if err = c.Watch(
 		&source.Kind{Type: &clusterv1.Cluster{}},
-		handler.EnqueueRequestsFromMapFunc(util.ClusterToInfrastructureMapFunc(infrav1exp.GroupVersion.WithKind("AzureManagedCluster"))),
+		handler.EnqueueRequestsFromMapFunc(util.ClusterToInfrastructureMapFunc(ctx, infrav1exp.GroupVersion.WithKind("AzureManagedCluster"), mgr.GetClient(), &infrav1exp.AzureManagedCluster{})),
```
Comment (Jont828) [inline exp/controllers/azuremanagedcluster_controller.go:89]: Just curious, what changed to the `util.ClusterToInfrastructureMapFunc()` to require multiple args now?

File: test/e2e/azure_failuredomains.go [hunk 6]
```diff
@@ -83,15 +82,9 @@ func AzureFailureDomainsSpec(ctx context.Context, inputGetter func() AzureFailur
 		// TODO: Find alternative when the number of zones is > 1 but doesn't equal to number of control plane machines.
 		if len(input.Cluster.Status.FailureDomains) == 3 {
 			By("Ensuring control planes are spread across availability zones.")
-			key := client.ObjectKeyFromObject(input.Cluster)
 			failureDomainsInput := framework.AssertControlPlaneFailureDomainsInput{
-				GetLister:  input.BootstrapClusterProxy.GetClient(),
-				ClusterKey: key,
-				ExpectedFailureDomains: map[string]int{
```
Comment (Jont828) [inline test/e2e/azure_failuredomains.go:90]: Just to double check, do we no longer need to have `ExpectedFailureDomains` here?

**Review by Jont828 — COMMENTED** [review]
Overall makes sense to me, just had a couple questions about the changes to CAPI 1.2.

File: exp/controllers/azuremanagedcluster_controller.go (hunk 5 above)
Comment (cpanato) [reply exp/controllers/azuremanagedcluster_controller.go:89]: here is the PR https://github.com/kubernetes-sigs/cluster-api/pull/6070

File: test/e2e/azure_failuredomains.go (hunk 6 above)
Comment (jackfrancis) [reply test/e2e/azure_failuredomains.go:90]: Correct, this is a documented change w/ 1.2.0. See this section:

- https://github.com/fabriziopandini/cluster-api/blob/main/docs/book/src/developer/providers/v1.1-to-v1.2.md#other

The "The AssertControlPlaneFailureDomains function in the E2E test framework has been modified..." part

**Comment by jackfrancis** [comment]
@CecileRobertMichon are there any remaining items to address before merging this?

> Would be good to add CAPI tests that were recently added to https://github.com/kubernetes-sigs/cluster-api/tree/main/test/e2e to test new v1.2 functionality to our CAPI test suite too: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/capi_test.go

The above I agree is valuable, I think we can do that as a follow-up so that we are in the pole position to ship 1.5.0 of capz from a build that has a good amount of capi 1.2.0 test validation.

**Comment by CecileRobertMichon** [comment]
> The above I agree is valuable, I think we can do that as a follow-up

can you please open an issue and mark it as help wanted?

/lgtm
/assign @mboersma 

Can the hold label be removed?

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

**Pull request merged**